File: cpp/n15_3sum.cpp
#include <vector>
#include <algorithm>
using std::vector;
using std::sort;
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int target2sum;
        vector<vector<int>> answer;
        sort(nums.begin(), nums.end());
        int left;
        int right;
        for(int i = 0; i < nums.size(); i++){
            if(i > 0){
                if(nums[i] == nums[i-1]){
                    continue;
                }
            }
            target2sum = -nums[i];
            left = i + 1;
            right = nums.size() - 1;
            while(left < right){
                if(nums[left] + nums[right] == target2sum){
                    vector<int> vec = {nums[i], nums[left], nums[right]};
                    if(left - 1 > i && nums[left] == nums[left - 1]){
                        left++;
                        right--;
                    }else{
                        answer.push_back(vec);
                        left++;
                        right--;
                    }
                }else if(nums[left] + nums[right] > target2sum){
                    right--;
                }else{
                    left++;
                }
            }
        }
        return answer;
    }
};

File: python/n15_3sum.py
from typing import *
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        l = len(nums)
        if l < 3:
            return []
        nums.sort()
        answer = []
        for i in range(l-2):
            if i > 0:
                if nums[i] == nums[i-1]:
                    continue
            j = i + 1
            k = l - 1
            while j < k:
                if nums[j] + nums[k] == -nums[i]:
                    if j-1 > i and nums[j] == nums[j-1]:
                        pass
                    else:
                        answer.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                elif nums[j] + nums[k] > -nums[i]:
                    k -= 1
                else:
                    j += 1
        return answer

if __name__ == "__main__":
    s = Solution()
    print(s.threeSum([-2,0,0,2,2]))